### PR TITLE
Correctly Handle Fields with Null Value for JSON Deserialization

### PR DESCRIPTION
**Pull request by chenpiaoping**
1. Correctly Handle Fields with Null Value for JSON Deserialization
2. Provides the interface to get # of ports in the subnet

### CODE REVIEW

File: services/port_manager/src/main/java/com/futurewei/alcor/portmanager/service/PortServiceImpl.java [hunk 1]
```diff
@@ -296,4 +295,9 @@ public RouterUpdateInfo updateL3Neighbors(String projectId, RouterUpdateInfo rou
 
         return routerUpdateInfo;
     }
+
+    @Override
+    public SubnetPortIds getSubnetPorts(String projectId, String subnetId) throws Exception {
+        return portRepository.getSubnetPortIds(subnetId);
+    }
```
Comment (cj-chung) [inline services/port_manager/src/main/java/com/futurewei/alcor/portmanager/service/PortServiceImpl.java:302]: @chenpiaoping This API just need to return # of ports in the subnet, you don't need to return their port ids. Subnet Manager will check this return value when the subnet is going to be deleted.

**Review by cj-chung — CHANGES_REQUESTED** [review]

**Comment by chenpiaoping** [comment]
#513 #505

File: services/port_manager/src/main/java/com/futurewei/alcor/portmanager/service/PortServiceImpl.java (hunk 1 above)
Comment (chenpiaoping) [reply services/port_manager/src/main/java/com/futurewei/alcor/portmanager/service/PortServiceImpl.java]: fixed, please review again.

File: services/port_manager/src/main/java/com/futurewei/alcor/portmanager/entity/SubnetPortIds.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+import java.util.Set;
+
+public class SubnetPortIds {
+    @JsonProperty("subnet_id")
+    private String subnetId;
+
+    @JsonProperty("port_ids")
+    private Set<String> portIds;
+
```
Comment (cj-chung) [inline services/port_manager/src/main/java/com/futurewei/alcor/portmanager/entity/SubnetPortIds.java:28]: @chenpiaoping Should include gateway port here and  distinguish it from a normal port? or add another field for gateway ports?

File: services/port_manager/src/main/java/com/futurewei/alcor/portmanager/controller/PortController.java [hunk 3]
```diff
@@ -221,4 +221,16 @@ public RouterUpdateInfo updateL3Neighbors(@PathVariable("project_id") String pro
         checkRouterSubnetUpdateInfo(routerUpdateInfo);
         return portService.updateL3Neighbors(projectId, routerUpdateInfo);
     }
+
+    @Rbac(resource ="port")
+    @GetMapping({"/project/{project_id}/subnet-port-number/{subnet_id}", "v4/{project_id}/subnet-port-number/{subnet_id}"})
+    @DurationStatistics
+    public int getSubnetPortNumber(@PathVariable("project_id") String projectId,
+                                             @PathVariable("subnet_id") String subnetId) throws Exception {
+        if (StringUtil.isNullOrEmpty(subnetId)) {
```
Comment (cj-chung) [inline services/port_manager/src/main/java/com/futurewei/alcor/portmanager/controller/PortController.java:230]: @chenpiaoping do you think function name `getSubnetPortCount` would be better? Same as the API name: `subnet-port-count`.

Comment (chenpiaoping) [reply services/port_manager/src/main/java/com/futurewei/alcor/portmanager/controller/PortController.java:230]: agree

File: services/port_manager/src/main/java/com/futurewei/alcor/portmanager/entity/SubnetPortIds.java (hunk 2 above)
Comment (chenpiaoping) [reply services/port_manager/src/main/java/com/futurewei/alcor/portmanager/entity/SubnetPortIds.java:28]: Right now portIds only contains non-gateway port.

**Comment by chenpiaoping** [comment]
@xieus I think this PR is almost ready, any other review suggestions?

**Comment by xieus** [comment]
> @xieus I think this PR is almost ready, any other review suggestions?

Let me quickly check.

File: services/port_manager/src/main/java/com/futurewei/alcor/portmanager/repo/PortRepository.java [hunk 4]
```diff
@@ -263,6 +264,33 @@ private String getNeighborCacheName(String suffix) {
         return NEIGHBOR_CACHE_NAME_PREFIX + suffix;
     }
 
+    private List<SubnetPortIds> getSubnetPortIds(List<PortEntity> portEntities) {
+        Map<String, SubnetPortIds> subnetPortIdsMap = new HashMap<>();
+        for (PortEntity portEntity: portEntities) {
+            if ("network:router_interface".equals(portEntity.getDeviceOwner())) {
```
Comment (xieus) [inline services/port_manager/src/main/java/com/futurewei/alcor/portmanager/repo/PortRepository.java:270]: A constant string would be better.

**Review by xieus — APPROVED** [review]
LGTM.

File: services/port_manager/src/main/java/com/futurewei/alcor/portmanager/entity/SubnetPortIds.java (hunk 2 above)
Comment (cj-chung) [reply services/port_manager/src/main/java/com/futurewei/alcor/portmanager/entity/SubnetPortIds.java:28]: ok

**Review by cj-chung — APPROVED** [review]
LGTM

File: services/port_manager/src/main/java/com/futurewei/alcor/portmanager/repo/PortRepository.java (hunk 4 above)
Comment (chenpiaoping) [reply services/port_manager/src/main/java/com/futurewei/alcor/portmanager/repo/PortRepository.java:270]: Fixed, another PR: https://github.com/futurewei-cloud/alcor/pull/519

**Pull request merged**